text/0249-v2-experiments.md

```diff
@@ -0,0 +1,127 @@
+installs aws-cdk-lib using npm or pip or any other package manager should be confident that there will be no intentional
+breaking changes in the 2.x line of releases for its lifetime.
+
+# README
+
+The following section will be removed from the CDK Readme:
+
+> Modules in the AWS Construct Library are designated Experimental while we build them; experimental modules may have
+> breaking API changes in any release. After a module is designated Stable, it adheres to semantic versioning, and only
+> major releases can have breaking changes. Each module's stability designation is available on its Overview page in the
+> AWS CDK API Reference. For more information, see Versioning in the CDK Developer Guide.
+
+The existing contributor guide does not mention experimental code, so the following will be added after the intro
+paragraph:
+
+> If you are contributing to a 2.x release, keep in mind that you may not introduce breaking changes to any public
+> facing API. Experimental modules are developed in a separate repository. [link]
+
+The new aws-cdk-previews repository will have the following Readme:
+
+> This repository contains experimental code related to the AWS Cloud Development Kit (AWS CDK). CDK is an open-source
+> software development framework to define cloud infrastructure in code and provision it through AWS CloudFormation.
+> Please refer to the main repository here [link]. [TODO...]
+
+The contributor guide can be copied from the main repository, since the toolchain for building the preview repository
+will be basically the same.
+
+# Design Summary
+
+- Upon release of v2.0 to GA, all experimental modules will be automatically marked as stable (in fact, the metadata in
+  the code regarding stability can be removed, since everything will be stable). This means that the opportunity for the
+  team and for the community to influence the shape of existing experimental APIs is between now and then. Once we go
+  GA, we have to live with our design decisions.
+- Any APIs that cannot be graduated with confidence in time for the v2 release will be deprecated in 1.x, removed from
+  the v2 branch, and moved to the preview repository. This option should be minimized, since some customers may have
+  deployed production stacks that depend on these APIs.
+- Create a new repository called aws-cdk-previews that will be the new home for experimental development and developer
+  preview releases. The packages in this repository will be vended in the same way that 1.x packages are vended today,
+  in a variety of languages, with metadata that draws a line between new experiments and more mature developer preview
+  packages. This repo will take a dependency on the core CDK package just like any other independent library. This
+  avoids the complications associated with experimenting on a fork or a branch of the main repository. The repository
```

Comment:
> avoids the complications associated with experimenting on a fork or a branch of the main repository

Why not experiment on `master` of the main repository, and just vend separate packages?

Reply:
I think it's a slippery slope. How would we guarantee there is no mixing of stable and experimental in a 2.x release? Also, why not separate repos? Why is that a problem?

Reply:
> How would we guarantee there is no mixing of stable and experimental in a 2.x release?

I feel like that's build-time validation that shouldn't be very hard to implement.

> Also, why not separate repos? Why is that a problem?

I can think of like 8 different reasons a separate repo can be a problem:

1. Customers will be confused which repo they need to contribute to.
2. There will be differences in working on the two repos (different build commands, code organization, linter configuration, etc.).
3. The second repo will be starved by the core team, because the main repo will take priority.
4. Changes cannot be coordinated between stable and experimental modules anymore.
5. We need to duplicate the entire release pipeline.
6. etc. (I'm sure there are more reasons if I thought about it more)

Reply:
@skinny85 I actually think

> 4. Changes cannot be coordinated between stable and experimental modules anymore.

 is a good thing, not a problem. Since the "stable" part is supposed to be... well... stable... then you shouldn't need coordination! If you need it, something smells bad.

Reply:
I just meant changing things in stable modules that are then used in experimental ones (of course, preserving backwards compatibility of the stable module). Today, you can do it in one commit; if we tear out the experimental modules to a separate repository, you have to first do it in the "stable" repository, wait at least a week for the release, and only then do the change in the "experimental" repository; which is another downside of the "separate repository" approach proposed here.